### DISCUSSION TITLE
Error converting data type nvarchar to int

### DISCUSSION THREAD

**Discussion by RebShira** [sqlalchemy/sqlalchemy#11543] [Usage Questions] [answered]
I am getting this error when I'm attempting to use the engine object to run a stored proc in MSSQL:

ProgrammingError at /helpdesk/submit
(pyodbc.ProgrammingError) ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Error converting data type nvarchar to int. (8114) (SQLExecDirectW)')
[SQL: EXEC [dbo].[sp_HelpdeskInitializeTicketUpdate] empid_it_assigned, importance_id]

Some details:

1. The issue has to do with the last parameter, importance_id. Unlike the first parameter, this one is an integer type. My code works fine with VARCHAR parameters.
2. The last 3 lines of code to execute before crashing are below. Note I am using the SQLAlchemy.text method...

```
com = sa.text(command)
with engine.begin() as conn:
   conn.execute(com, parameters)
```

**Comment by gordthompson** [comment]
You are apparently passing a string for `importance_id` that SQL Server cannot convert to an integer. What does `repr(parameters)` look like?

**Reply by RebShira** [reply]
Thanks for taking the time to help. It doesn't make a difference whether I convert the value to an integer Something like:

`EXEC SPROC sproc1 {"string1": "val1", "string2: "val2", "importance_id": intval}`

The error was the same, whether or not importance_id's value was surrounded by quotation marks in what the SQLAlchemy engine passed.

**Reply by gordthompson** [reply]
So what does `repr(intval)` look like?

**Reply by gordthompson** [reply, marked as answer]
Passing a string value to an int parameter is not a problem *per se*, it just needs to be something that SQL Server can convert to int.

```python
import sqlalchemy as sa

engine = sa.create_engine("mssql+pyodbc://scott:tiger^5HHH@msodbcsql18_199")

with engine.begin() as conn:
    conn.exec_driver_sql("DROP PROCEDURE IF EXISTS sproc1")
    conn.exec_driver_sql("""\
CREATE PROCEDURE sproc1
@some_text nvarchar(50),
@importance_id int
AS
BEGIN
    SET NOCOUNT ON;
    -- this sp doesn't actually do anything
END
    """)

with engine.begin() as conn:
    stmt = sa.text("EXEC sproc1 :some_text, :importance_id")
    conn.execute(stmt, {"some_text": "val1", "importance_id": 123})
    # no error
    conn.execute(stmt, {"some_text": "val1", "importance_id": "123"})
    # also no error
    conn.execute(stmt, {"some_text": "val1", "importance_id": "1,000"})
    # Error converting data type nvarchar to int. (8114) (SQLExecDirectW)
```

**Reply by RebShira** [reply]
I could define the parameter as a VARCHAR, and have the procedure convert it into an integer. It's obviously messy, but it will work...
